File: python/xpath-helper/xpath_helper/filter.py
import functools
from typing import List, Optional

"""
The following Filter classes provide a simple, chainable and decomposable api
to create XPath filter expression
"""

"""
Shortcut to design any attribute name
"""
ANY_ATTRIBUTE = "*"


"""
XPath Filter containing a valid expression.
"""


class ValidExpressionFilter:
    sb: List[str] = []

    def __init__(self, current_path: Optional[List[str]]=None):
        """Creates an instance of ValidExpressionFilter.

        Args:
            currentPath (list[string]): Current filter path
        """
        if (current_path != None):
            self.sb = current_path

    def and_operator(self, *filters: 'EmptyFilter') -> 'ValidExpressionFilter':
        """Adds one or more filter expression to the current one with the AND logical operator.

        Args:
            filters (list[Filter]): List of filters List of filters that will be added with the AND logical operator

        Returns:
            ValidExpressionFilter: a new instance of ValidExpressionFilter with the newly formed expression.
        """
        expression = ""
        if (len(self.sb) != 0):
            expression += " and "

        expression += "("
        expression += functools.reduce(lambda acc, filter: acc + add_openrand(
            filter, " and ", filter == filters[-1]), filters, "")
        expression += ")"
        return ValidExpressionFilter(self.sb + [expression])

    def or_operator(self, *filters: 'EmptyFilter') -> 'ValidExpressionFilter':
        """Adds one or more filter expression to the current one with the OR logical operator.

        Args:
            filters (list[Filter]): List of filters that will be added with the OR logical operator

        Returns:
            ValidExpressionFilter: a new instance of ValidExpressionFilter with the newly formed expression.
        """
        expression = ""
        if (len(self.sb) != 0):
            expression += " or "

        expression += "("
        expression += functools.reduce(lambda acc, filter: acc + add_openrand(
            filter, " or ", filter == filters[-1]), filters, "")
        expression += ")"
        return ValidExpressionFilter(self.sb + [expression])

    def __str__(self):
        """Returns the Filter as a valid XPath filter expression.

        Returns:
            str: the string of the corresponding XPath filter
        """
        return "".join(self.sb)

    def empty(self):
        """Empties the current path.
        """
        self.sb = []

    def is_empty(self) -> bool:
        """Returns true if filter is empty.

        Returns:
            bool: true if filter is empty
        """
        return len(self.sb) == 0


"""
Empty XPath filter.
"""


class EmptyFilter(ValidExpressionFilter):
    def __init__(self, currentPath=None):
        """Creates an instance of Filter.

        Args:
            currentPath (list[string]): Current filter path
        """
        super().__init__(currentPath)

    def has_attribute(self, attribute: str) -> ValidExpressionFilter:
        """Selects the nodes that have the attribute <code>attribute</code>.

        Args:
            attribute (str): attribute name

        Returns:
            ValidExpressionFilter: a new instance of ValidExpressionFilter with the newly formed expression.
        """
        return ValidExpressionFilter(self.sb + ["@" + attribute])

    def attribute_contains(self, attribute: str, value: str) -> ValidExpressionFilter:
        """Selects the nodes with the attribute <code>attribute</code> containing the value <code><value</code>.

        Args:
            attribute (str): attribute name
            value (str): attribute value

        Returns:
            ValidExpressionFilter: a new instance of ValidExpressionFilter with the newly formed expression.
        """
        return ValidExpressionFilter(self.sb + ["contains(@" + attribute + ", " + replace_apostrophes(value) + ")",
                                                ])

    def attribute_equals(self, attribute: str, value: str) -> ValidExpressionFilter:
        """Selects the nodes with the attribute <code>attribute</code>, whose value equals <code><value</code>.

        Args:
            attribute (str): attribute name
            value (str | int | float): attribute value

        Returns:
            ValidExpressionFilter: a new instance of ValidExpressionFilter with the newly formed expression.
        """
        return ValidExpressionFilter(self.sb + ["@" + attribute + "=" + replace_apostrophes(value) + "",
                                                ])

    def attribute_not_equals(self, attribute: str, value: str) -> ValidExpressionFilter:
        """Selects the nodes with the attribute <code>attribute</code>, whose value doesn't equal <code><value</code>.

        Args:
            attribute (str): attribute name
            value (str | int | float): attribute value

        Returns:
            ValidExpressionFilter: a new instance of ValidExpressionFilter with the newly formed expression.
        """
        return ValidExpressionFilter(self.sb + ["@" + attribute + "!=" + replace_apostrophes(value)])

    def attribute_less_than(self, attribute: str, value: str) -> ValidExpressionFilter:
        """Selects the nodes with the attribute <code>attribute</code>, whose value is less than <code><value<code>.

        Args:
            attribute (str): attribute name
            value (int | float): attribute value

        Returns:
            ValidExpressionFilter: a new instance of ValidExpressionFilter with the newly formed expression.
        """
        return ValidExpressionFilter(self.sb + ["@" + attribute + "<" + str(value)])

    def attribute_less_than_or_equal_to(self, attribute: str, value: str) -> ValidExpressionFilter:
        """Selects the nodes with the attribute <code>attribute</code>, whose value is less than or equal to <code><value</code>.

        Args:
            attribute (str): attribute name
            value (int | float): attribute value

        Returns:
            ValidExpressionFilter: a new instance of ValidExpressionFilter with the newly formed expression.
        """
        return ValidExpressionFilter(self.sb + ["@" + attribute + "<=" + str(value)])

    def attribute_greater_than(self, attribute: str, value: str) -> ValidExpressionFilter:
        """Selects the nodes with the attribute <code>attribute</code>, whose value is greater than <code><value</code>.

        Args:
            attribute (str): attribute name
            value (int | float): attribute value

        Returns:
            ValidExpressionFilter: a new instance of ValidExpressionFilter with the newly formed expression.
        """
        return ValidExpressionFilter(self.sb + ["@" + attribute + ">" + str(value)])

    def attribute_greater_than_or_equal_to(self, attribute: str, value: str) -> ValidExpressionFilter:
        """Selects the nodes with the attribute <code>attribute</code>, whose value is greater than or equal to <code><value</code>.

        Args:
            attribute (str): attribute name
            value (int | float): attribute value

        Returns:
            ValidExpressionFilter: a new instance of ValidExpressionFilter with the newly formed expression.
        """
        return ValidExpressionFilter(self.sb + ["@" + attribute + ">=" + str(value)])

    def value_contains(self, value: str) -> ValidExpressionFilter:
        """Selects the nodes containing the value <code><value</code>.

        Args:
            value (str): value

        Returns:
            ValidExpressionFilter: a new instance of ValidExpressionFilter with the newly formed expression.
        """
        return ValidExpressionFilter(self.sb + ["text()[contains(., " + replace_apostrophes(value) + ")]"])

    def value_equals(self, value: str) -> ValidExpressionFilter:
        """Selects the nodes whose value equals <code><value</code>.

        Args:
            value (str | int | float): value

        Returns:
            ValidExpressionFilter: a new instance of ValidExpressionFilter with the newly formed expression.
        """
        return ValidExpressionFilter(self.sb + ["text() = " + replace_apostrophes(value)])

    def value_not_equals(self, value: str) -> ValidExpressionFilter:
        """Selects the nodes with whose value doesn't equal <code><value</code>.

        Args:
            value (str | int | float): value

        Returns:
            ValidExpressionFilter: a new instance of ValidExpressionFilter with the newly formed expression.
        """
        return ValidExpressionFilter(self.sb + ["text() !=" + replace_apostrophes(value)])

    def value_less_than(self, value: str) -> ValidExpressionFilter:
        """Selects the nodes whose value is less than <code><value</code>.

        Args:
            value (int | float): value

        Returns:
            ValidExpressionFilter: a new instance of ValidExpressionFilter with the newly formed expression.
        """
        return ValidExpressionFilter(self.sb + ["text() <" + str(value)])

    def value_less_than_or_equal_to(self, value: str) -> ValidExpressionFilter:
        """Selects the nodes whose value is less than or equal to <code><value</code>.

        Args:
            value (int | float): value

        Returns:
            ValidExpressionFilter: a new instance of ValidExpressionFilter with the newly formed expression.
        """
        return ValidExpressionFilter(self.sb + ["text() <=" + str(value)])

    def value_greater_than(self, value: str) -> ValidExpressionFilter:
        """Selects the nodes  whose value is greater than <code><value</code>.

        Args:
            value (int | float): value

        Returns:
            ValidExpressionFilter: a new instance of ValidExpressionFilter with the newly formed expression.
        """
        return ValidExpressionFilter(self.sb + ["text() >" + str(value)])

    def value_greater_than_or_equal_to(self, value: str) -> ValidExpressionFilter:
        """Selects the nodes whose value is greater than or equal to <code><value</code>.

        Args:
            value (int | float): value

        Returns:
            ValidExpressionFilter: a new instance of ValidExpressionFilter with the newly formed expression.
        """
        return ValidExpressionFilter(self.sb + ["text() >=" + str(value)])

    def get(self, index: int):
        """Selects the node element who is positioned at the <code>index</code> position in its parent children list.

        Args:
            index (int): index of the element in the list

        Returns:
            ValidExpressionFilter: a new instance of ValidExpressionFilter with the newly formed expression.
        """
        return ValidExpressionFilter(self.sb + [str(index)])

    def get_first(self):
        """Selects the node element who is positioned first in its parent children list.

        Returns:
            ValidExpressionFilter: a new instance of ValidExpressionFilter with the newly formed expression.
        """
        return ValidExpressionFilter(self.sb + ["1"])

    def get_last(self):
        """Selects the node element who is positioned last in its parent children list.

        Returns:
            ValidExpressionFilter: a new instance of ValidExpressionFilter with the newly formed expression.
        """
        return ValidExpressionFilter(self.sb + ["last()"])

    def not_operator(self, filter: ValidExpressionFilter):
        """Reverses the filter <code>filter</code>. Returns true when the filter returns false and true when the filter returns false.

        Args:
            filter ([type]): [description]

        Returns:
            ValidExpressionFilter: a new instance of ValidExpressionFilter with the newly formed expression.
        """
        return ValidExpressionFilter(self.sb + ["not( " + add_openrand(filter) + " )"])


def add_openrand(filter: EmptyFilter, separator="", is_last: Optional[bool] = True) -> str:
    """Adds operand between filters

    Args:
        filter (Filter): a filter
        separator (str, optional): " and " of " or " separator. Defaults to "".
        isLast (bool, optional): True if it is the last operand. Defaults to True.

    Returns:
        str: opperand with and / or separator
    """
    suffix: str = ""
    if filter and not filter.is_empty():
        expression = str(filter)
        suffix = expression
        if not is_last:
            suffix += separator
    return suffix


def replace_apostrophes(input: str) -> str:
    """Treats the presence of apostrophes so it doesn't break the XPath filter expression.

    Args:
        input (str | int): input

    Returns:
        str: XPath filter expression with apostrophes handled.
    """
    if not isinstance(input, str):
        return str(input)

    if "'" in input:
        prefix: str = ""
        elements = input.split("'")
        output = "concat("
        for s in elements:
            output += prefix + "'" + s + "'"
            prefix = ',"\'",'

        output += ")"
        return output

    else:
        return "'" + input + "'"

File: javascript/xpath-helper/src/filter.ts
/**
 * The following EmptyFilter classes provide a simple, chainable and decomposable api
 * to create XPath filter expression
 */

/**
 * Interface providing the minimal methods that must implement an XPath EmptyFilter.
 * @export
 * @interface IFilter
 */
export interface IFilter {
  toString(): string;
  isEmpty(): boolean;
}

/**
 * XPath Filter containing a valid expression.
 * @class ValidExpressionFilter
 * @implements {IFilter}
 */
export class ValidExpressionFilter implements IFilter {
  /**
   * Array of string containing the different part of the filter expression.
   * @protected
   * @type {string[]}
   * @memberof ValidExpressionFilter
   */
  protected sb: string[];

  /**
   * Creates an instance of ValidExpressionFilter.
   * @param {string[]} [currentPath] current path
   * @memberof ValidExpressionFilter
   */
  constructor(currentPath?: string[]) {
    if (currentPath) {
      this.sb = currentPath;
    } else {
      this.sb = new Array<string>();
    }
  }

  /**
   * Adds one or more filter expression to the current one with the AND logical operator.
   * @param {IFilter[]} filters List of filters that will be added with the AND logical operator
   * @returns {ValidExpressionFilter} a new instance of ValidExpressionFilter with the newly formed expression
   * @memberof ValidExpressionFilter
   */
  and(...filters: IFilter[]) {
    let expression = "";
    if (this.sb.length !== 0) {
      expression += " and ";
    }
    expression += "(";
    filters.map(
      (filter, i) =>
        (expression += addOpenrand(filter, " and ", i === filters.length - 1))
    );
    return new ValidExpressionFilter([...this.sb, expression, ")"]);
  }

  /**
   * Adds one or more filter expression to the current one with the OR logical operator.
   * @param {IFilter[]} filters List of filters that will be added with the OR logical operator
   * @returns {ValidExpressionFilter} a new instance of ValidExpressionFilter with the newly formed expression
   * @memberof ValidExpressionFilter
   */
  or(...filters: IFilter[]) {
    let expression = "";
    if (this.sb.length !== 0) {
      expression += " or ";
    }
    expression += "(";
    filters.map(
      (filter, i) =>
        (expression += addOpenrand(filter, " or ", i === filters.length - 1))
    );
    return new ValidExpressionFilter([...this.sb, expression, ")"]);
  }

  /**
   * Returns the EmptyFilter as a valid XPath filter expression.
   * @returns {string} a valid XPath filter expression.
   * @memberof ValidExpressionFilter
   */
  public toString(): string {
    return this.sb.join("");
  }

  /**
   * Empties the current path.
   * @memberof ValidExpressionFilter
   */
  public empty(): void {
    this.sb = new Array<string>();
  }

  /**
   * Returns true if filter is empty.
   * @returns {boolean} true if filter is empty.
   * @memberof ValidExpressionFilter
   */
  public isEmpty(): boolean {
    return this.sb.length === 0;
  }
}

/**
 * Empty XPath filter.
 * @export
 * @class EmptyFilter
 * @extends {ValidExpressionFilter}
 */
export class EmptyFilter extends ValidExpressionFilter {
  /**
   * Shortcut to design any attribute name
   * @static
   * @type {string}
   * @memberof EmptyFilter
   */
  public static ANY_ATTRIBUTE: string = "*";

  /**
   * Creates an instance of EmptyFilter.
   * @param {string[] } [currentPath] current path
   * @memberof EmptyFilter
   */
  constructor(currentPath?: string[]) {
    super(currentPath);
  }

  /**
   * Selects the nodes that have the attribute ``attribute``.
   * @param {string} attribute - attribute name
   * @returns {ValidExpressionFilter} a new instance of ValidExpressionFilter with the newly formed expression.
   * @memberof EmptyFilter
   */
  hasAttribute(attribute: string) {
    return new ValidExpressionFilter([...this.sb, "@" + attribute]);
  }

  /**
   * Selects the nodes with the attribute ``attribute`` containing the value ``value``.
   * @param attribute
   * @param value
   * @returns {ValidExpressionFilter} a new instance of ValidExpressionFilter with the newly formed expression.
   */
  attributeContains(attribute: string, value: string) {
    return new ValidExpressionFilter([
      ...this.sb,
      "contains(@" + attribute + ", " + replaceApostrophes(value) + ")"
    ]);
  }

  /**
   * Selects the nodes with the attribute ``attribute``, whose value equals ``value``.
   * @param {string} attribute - attribute name
   * @param {(string | number)} value - value
   * @returns {ValidExpressionFilter} a new instance of ValidExpressionFilter with the newly formed expression.
   * @memberof EmptyFilter
   */
  attributeEquals(attribute: string, value: string | number) {
    return new ValidExpressionFilter([
      ...this.sb,
      "@" + attribute + "=" + replaceApostrophes(value) + ""
    ]);
  }

  /**
   * Selects the nodes with the attribute ``attribute``, whose value doesn't equal ``value``.
   * @param {string} attribute - attribute name
   * @param {(string | number)} value - value
   * @returns {ValidExpressionFilter} a new instance of ValidExpressionFilter with the newly formed expression.
   * @memberof EmptyFilter
   */
  attributeNotEquals(attribute: string, value: string | number) {
    return new ValidExpressionFilter([
      ...this.sb,
      "@" + attribute + "!=" + replaceApostrophes(value) + ""
    ]);
  }

  /**
   * Selects the nodes with the attribute ``attribute``, whose value is less than ``value``.
   * @param {string} attribute - attribute name
   * @param {number} value - value
   * @returns {ValidExpressionFilter} a new instance of ValidExpressionFilter with the newly formed expression.
   * @memberof EmptyFilter
   */
  attributeLessThan(attribute: string, value: number) {
    return new ValidExpressionFilter([...this.sb, "@" + attribute + "<" + value]);
  }

  /**
   * Selects the nodes with the attribute ``attribute``, whose value is less than or equal to ``value``.
   * @param {string} attribute - attribute name
   * @param {number} value - value
   * @returns {ValidExpressionFilter} a new instance of ValidExpressionFilter with the newly formed expression.
   * @memberof EmptyFilter
   */
  attributeLessThanOrEqualTo(attribute: string, value: number) {
    return new ValidExpressionFilter([...this.sb, "@" + attribute + "<=" + value]);
  }

  /**
   * Selects the nodes with the attribute ``attribute``, whose value is greater than ``value``.
   * @param {string} attribute - attribute name
   * @param {number} value - value
   * @returns {ValidExpressionFilter} a new instance of ValidExpressionFilter with the newly formed expression.
   * @memberof EmptyFilter
   */
  attributeGreaterThan(attribute: string, value: number) {
    return new ValidExpressionFilter([...this.sb, "@" + attribute + ">" + value]);
  }
  /**
   * Selects the nodes with the attribute ``attribute``, whose value is greater than or equal to ``value``.
   * @param {string} attribute - attribute name
   * @param {number} value - value
   * @returns {ValidExpressionFilter} a new instance of ValidExpressionFilter with the newly formed expression.
   * @memberof EmptyFilter
   */
  attributeGreaterThanOrEqualTo(attribute: string, value: number) {
    return new ValidExpressionFilter([...this.sb, "@" + attribute + ">=" + value]);
  }
  /**
   * Selects the nodes containing the value ``value``.
   * @param {string} value - node value
   * @returns {ValidExpressionFilter} a new instance of ValidExpressionFilter with the newly formed expression.
   * @memberof EmptyFilter
   */
  valueContains(value: string) {
    return new ValidExpressionFilter([
      ...this.sb,
      "text()[contains(., " + replaceApostrophes(value) + ")]"
    ]);
  }

  /**
   * Selects the nodes whose value equals ``value``.
   * @param {(string | number)} value - node value
   * @returns {ValidExpressionFilter} a new instance of ValidExpressionFilter with the newly formed expression.
   * @memberof EmptyFilter
   */
  valueEquals(value: string | number) {
    return new ValidExpressionFilter([
      ...this.sb,
      "text() = " + replaceApostrophes(value) + ""
    ]);
  }

  /**
   * Selects the nodes with whose value doesn't equal ``value``.
   * @param {(string | number)} value - node value
   * @returns {ValidExpressionFilter} a new instance of ValidExpressionFilter with the newly formed expression.
   * @memberof EmptyFilter
   */
  valueNotEquals(value: string | number) {
    return new ValidExpressionFilter([
      ...this.sb,
      "text() !=" + replaceApostrophes(value) + ""
    ]);
  }

  /**
   * Selects the nodes whose value is less than ``value``.
   * @param {number} value - node value
   * @returns {ValidExpressionFilter} a new instance of ValidExpressionFilter with the newly formed expression.
   * @memberof EmptyFilter
   */
  valueLessThan(value: number) {
    return new ValidExpressionFilter([...this.sb, "text() <" + value]);
  }

  /**
   * Selects the nodes whose value is less than or equal to ``value``.
   * @param {number} value - node value
   * @returns {ValidExpressionFilter} a new instance of ValidExpressionFilter with the newly formed expression.
   * @memberof EmptyFilter
   */
  valueLessThanOrEqualTo(value: number) {
    return new ValidExpressionFilter([...this.sb, "text() <=" + value]);
  }

  /**
   * Selects the nodes whose value is greater than ``value``.
   * @param {number} value - node value
   * @returns {ValidExpressionFilter} a new instance of ValidExpressionFilter with the newly formed expression.
   * @memberof EmptyFilter
   */
  valueGreaterThan(value: number) {
    return new ValidExpressionFilter([...this.sb, "text() >" + value]);
  }

  /**
   * Selects the nodes whose value is greater than or equal to ``value``.
   * @param {number} value - node value
   * @returns {ValidExpressionFilter} a new instance of ValidExpressionFilter with the newly formed expression.
   * @memberof EmptyFilter
   */
  valueGreaterThanOrEqualTo(value: number) {
    return new ValidExpressionFilter([...this.sb, "text() >=" + value]);
  }

  /**
   * Selects the node element who is positioned at the ``index`` position in its parent children list.
   * @param {number} index - index value
   * @returns {ValidExpressionFilter} a new instance of ValidExpressionFilter with the newly formed expression.
   * @memberof EmptyFilter
   */
  get(index: number) {
    return new ValidExpressionFilter([...this.sb, "" + index]);
  }

  /**
   * Selects the node element who is positioned first in its parent children list.
   * @returns {ValidExpressionFilter} a new instance of ValidExpressionFilter with the newly formed expression.
   * @memberof EmptyFilter
   */
  getFirst() {
    return new ValidExpressionFilter([...this.sb, "1"]);
  }

  /**
   * Selects the node element who is positioned last in its parent children list.
   * @returns {ValidExpressionFilter} a new instance of ValidExpressionFilter with the newly formed expression.
   * @memberof EmptyFilter
   */
  getLast() {
    return new ValidExpressionFilter([...this.sb, "last()"]);
  }

  /**
   * Reverses the filter ``filter``. Returns true when the filter returns false and true when the filter returns false.
   * @param {IFilter} filter - filter object
   * @returns {ValidExpressionFilter} a new instance of ValidExpressionFilter with the newly formed expression.
   * @memberof EmptyFilter
   */
  not(filter: IFilter) {
    return new ValidExpressionFilter([...this.sb, "not( " + addOpenrand(filter) + " )"]);
  }
}

/**
 * Adds operand between filters
 * @param {IFilter} filter - filter object
 * @param {string} [separator=""] - separator to add after expression 
 * @param {boolean} [isLast=true] - true if this operand is the last of the conditional expression
 * @returns {string}
 */
function addOpenrand(filter: IFilter, separator = "", isLast = true): string {
  let suffix = "";
  if (filter && !filter.isEmpty()) {
    const expression = filter.toString();
    suffix = expression;
    if (!isLast) {
      suffix += separator;
    }
  }
  return suffix;
}

/**
 * Treats the presence of apostrophes so it doesn't break the XPath filter expression.
 * @param {(string | number)} input - input value
 * @returns {string} XPath filter expression with apostrophes handled.
 */
function replaceApostrophes(input: string | number) {
  if (typeof input === "number") {
    return input;
  }
  if (input.includes("'")) {
    let prefix: string = "";
    const elements: string[] = input.split("'");
    let output: string = "concat(";

    for (const s of elements) {
      output += prefix + "'" + s + "'";
      prefix = ',"\'",';
    }

    output += ")";
    return output;
  } else {
    return "'" + input + "'";
  }
}
